src/bar.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt;
+
+use crate::pci_config::{validate_dword_alignment, Error as ConfigSpaceAccessError};
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+/// Error type for PCI Base Address Register accesses.
+pub enum Error {
+    /// Invalid access in the configuration space.
+    Access(ConfigSpaceAccessError),
+    /// Memory region defined by the BAR address and lenght is invalid.
+    BarAddressInvalid(u64, u64),
+}
+
+impl std::error::Error for Error {}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use Error::*;
+
+        match self {
+            Access(e) => write!(f, "config space access: {}", e),
+            BarAddressInvalid(addr, len) => {
+                write!(f, "out of bounds BAR addr {} len {}", addr, len)
+            }
+        }
+    }
+}
+
+impl From<ConfigSpaceAccessError> for Error {
+    fn from(e: ConfigSpaceAccessError) -> Self {
+        Error::Access(e)
+    }
+}
+
+#[derive(Copy, Clone)]
+/// Enum representing the "prefetchable" state masks of BARs (bit 3 in the register).
+pub enum PciBarPrefetchable {
+    /// Non-prefetchable BARs have bit 3 set to 0.
+    NotPrefetchable = 0,
+    /// Prefetchable BARs have bit 3 set to 1.
+    Prefetchable = 0x08,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+/// PCI BAR memory region abstraction.
+pub enum PciBarRegion {
+    /// Represents a 32-bit IO space region.
+    Io {
+        /// Start address of the 32-bit IO space region.
+        addr: u32,
+        /// Length of the region, in bytes.
+        len: u32,
+    },
+    /// Represents a 32-bit memory mapped region.
+    Memory32Bit {
+        /// Start address of the 32-bit memory mapped region.
+        addr: u32,
+        /// Length of the region, in bytes.
+        len: u32,
+    },
+    /// Represents a 64-bit memory mapped region.
+    Memory64Bit {
+        /// Start address of the 64-bit memory mapped region.
+        addr: u64,
+        /// Length of the region, in bytes.
+        len: u64,
+    },
+}
+
+impl PciBarRegion {
+    /// Describe an IO space BAR region.
+    pub fn new_io_region(addr: u32, len: u32) -> Result<Self> {
+        addr.checked_add(len)
+            .ok_or(Error::BarAddressInvalid(addr.into(), len.into()))?;
```

Comment:
`ok_or` should be replaced everywhere with `ok_or_else` because `ok_or` eagerly executes what is passed as the lambda function.